### PR TITLE
Adaptive site: hydrate islands according to their priority

### PR DESCRIPTION
**Pull request by sndrs**
_depends on #8970..._

## What does this change?

- uses the scheduler to hydrate islands according to their `priority` prop

## Why?

- ensures higher priority island are hydrated before lower priority tasks are run

### CODE REVIEW

**Review by mxdvl — COMMENTED** [review]
Nice and clean!

File: dotcom-rendering/src/client/islands/initHydration.ts [hunk 1]
```diff
@@ -45,7 +46,12 @@ export const initHydration = async (
 	switch (deferUntil) {
 		case 'idle': {
 			whenIdle(() => {
-				void doHydration(name, props, element, emotionCache, config);
+				void schedule(
+					name,
+					() =>
+						doHydration(name, props, element, emotionCache, config),
+					{ priority: 'feature' },
```
[the remark this answers is not in the document]
Comment (sndrs) [reply dotcom-rendering/src/client/islands/initHydration.ts:53]: no they're not, good spot!

[the remark this answers is not in the document]
Comment (sndrs) [reply dotcom-rendering/src/client/islands/initHydration.ts]: this should be fixed now, thanks!

File: dotcom-rendering/src/client/islands/initHydration.ts [hunk 2]
```diff
@@ -38,51 +41,39 @@ export const initHydration = async (
...
-					props,
-					element,
-					emotionCache,
-					config,
-				).then(() => {
-					targetElement.dispatchEvent(new MouseEvent('click'));
-				});
+			onInteraction(element, async (targetElement) => {
+				await scheduleHydration();
+				targetElement.dispatchEvent(new MouseEvent('click'));
```
[the remark this answers is not in the document]
Comment (sndrs) [reply dotcom-rendering/src/client/islands/initHydration.ts:70]: i think it's just much clearer, there's less language scaffolding around the logic - do you prefer the previous `then`able stuff?

[the remark this answers is not in the document]
Comment (sndrs) [reply dotcom-rendering/src/client/islands/initHydration.ts:70]: is also avoids needing to create a new function in the `then`, but i'm not sure that's a _huge_ win...

File: dotcom-rendering/src/client/islands/initHydration.ts [hunk 3]
```diff
@@ -38,51 +41,39 @@ export const initHydration = async (
...
 			});
 			return;
 		}
 		case 'hash': {
 			if (window.location.hash.includes(name) || hasLightboxHash(name)) {
-				void doHydration(name, props, element, emotionCache, config);
+				return scheduleHydration();
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/client/islands/initHydration.ts:76]: Indeed, this means that when `initHydration` is called on a `<Island defer={{until: 'hash'}}>` and the page URL contains a hash matching its name, the promise resolves only once the Island is fully hydrated.

This is something we should have caught in #8602

File: dotcom-rendering/src/client/islands/onInteraction.ts [hunk 4]
```diff
@@ -7,13 +7,13 @@
  */
 export const onInteraction = (
 	element: HTMLElement,
-	callback: (e: HTMLElement) => void,
+	callback: (e: HTMLElement) => Promise<void>,
```
Comment (mxdvl) [inline dotcom-rendering/src/client/islands/onInteraction.ts:10]: Is this change stricly necessary? If so, should it be `PromiseLike<void>`?

File: dotcom-rendering/src/client/islands/whenVisible.ts [hunk 5]
```diff
@@ -13,7 +13,7 @@ type WhenVisibleOptions = {
  */
 export const whenVisible = (
 	element: HTMLElement,
-	callback: () => void,
+	callback: () => Promise<void>,
```
Comment (mxdvl) [inline dotcom-rendering/src/client/islands/whenVisible.ts:16]: Same as above

**Review by mxdvl — COMMENTED** [review]
Maybe we should keep the `whenVisible` and `onInteraction` callback signatures as they were?

**Pull request merged**